Readme.md
# mdPushButton Library

An [Arduino](https://www.arduino.cc/) library capable of returning the number of consecutive 
presses of a push button made in quick succession or reporting if the 
button was held down for a long time.

## Version

The original version of this library was made available elsewhere under 
the name [mdButton.zip](https://sigmdel.ca/michel/program/esp8266/arduino/switch_debouncing_en.html#downloads). 
Version 0.1.1 (2020-12-04) adopted a better name, improved the class 
constructor to handle active `HIGH` buttons with  internal 
pull-down resistors if available and added callback functions.

Version 0.1.2 (2021-01-04) 
1. Added button release debounce on long button presses.
2. Removed the `_checkInterval` and `setCheckInterval`. The functionality was not implemented.
3. Improved the debugging functions in the `mdPushButton` class.
4. Added the `push_button_tune.ino` sketch.


## mdPushButton Class Constructor

An instance of the mdPushButton class must be create for each
push button. The GPIO pin connected to one terminal of the button must 
be specified in the constructor.

    mdPushButton(uint8_t pin, uint8_t active = LOW, bool useInternalPullResistor = true);

It is assumed that when pressed, the signal from the button is
brought `LOW`, but that can be changed by `active` to `HIGH`. It is
also assumed that an internal pull-up resistor will be used to 
keep the GPIO pin `HIGH` when the push button is not activated. If
an external pull-up resistor is in place, `useInternalPullResistor`
can be set to `false`.

If the push button brings the `active HIGH` and `useInternalPullResistor` 
is `true` then an internal pull-down resistor will be activated if
that is supported.

## Polling

The state of the push button is updated when the function 

    int status(void)
    
is invoked. This function must be called at regular intervals which
is usually done in the `loop()` function of the sketch. The value 
returned by the function indicates the state of the push button:

    -1   button held down for a long time
     0   button is not pressed
     1+  number of consecutive times the button has been pressed

## Delays

A debounce delay is used when the button is pressed and when
the button is released. These attributes may be set with the 
functions

    void setDebouncePressTime(uint16_t value);
    void setDebounceReleaseTime(uint16_t value);

The time values are in milliseconds (ms).

The library waits a few milliseconds after debouncing a button
release to see if there is a subsequent button press. This wait
period is set with the function

    void setMultiClickTime(uint16_t value);
    
If this time is too short then instead of reporting a number of
consecutive button presses, the library will report distinct single
button presses. If the time is too long, the library will seem 
unresponsive.

Just how long the button must be held for a long press is set
with the function 

    void setHoldTime(uint16_t value);
   
  
The default for all these time values are as follows:

	#define DEFAULT_DEBOUNCE_PRESS_TIME      15  // delay to debounce the make part of the signal
	#define DEFAULT_DEBOUNCE_RELEASE_TIME    30  // delay to debounce the break part of the signal
	#define DEFAULT_MULTI_CLICK_TIME        400  // if 0, does not check for multiple button clicks
	#define DEFAULT_HOLD_TIME              2000  // minimum time of button press for mdButton.status() to return a -1 (long button press)

Again, all times are in milliseconds.

The physical properties of push buttons can be considerable even when comparing two switches 
of the same type made by the same producer. The default time values are conservative. Use the 
`push_button_tune.ino` sketch in the `examples` directory to test other delay values with a given switch.


## Callback Functions

Instead of testing the returned value of `status()` to see if the button 
has been pressed, a callback function can be specified. The simplest
form of a call back will be `void` function with a single integer
parameter which will be the returned number of button presses.

    void buttonPressed(int clicks)

This callback is assigned with the 'OnButtonPressed()` method.

    void OnButtonClicked(buttonPressed);

It is also possible to install a callback function with two integer
parameters. The first will be the GPIO pin number of the button,
the second the number of consecutive button presses. This is useful
when a single handler for multiple buttons is used.

    void buttonPressed2(uint8_t pin, int clicks)
    
This callback is assigned with an overloaded version of the 
'OnButtonPressed()` method.

    void OnButtonClicked(buttonPressed2);
    

# Examples

Hopefully, the examples illustrate how to use this simple library. 
The first, `push_button_basic.ino` uses the returned value from the
`status()` function directly and must therefore correctly handle a 
returned `0` which indicates that no button has been pressed. 

The second example, `push_button_callback.ino` is similar but uses a 
callback function which will never have to handle a no button pressed 
state. Note that `status()` must nevertheless be called in the 
sketch `loop()`. 


The third example, `push_button_callback2.ino` shows how a single callback can 
be used with multiple buttons.

The `push_button_tune.ino` sketch is used to find the appropriate debounce time
values and so on for a given switch. Connect the switch to an appropriate micro-controller 
development board with a supported serial port. Enter `help` in the serial
monitor for a list of commands.


# Credits

This an implementation of a debounce algorithm by Jack G. Gannsle 
described in 
[A Guide to Debouncing](http://www.eng.utah.edu/~cs5780/debouncing.pdf) 
(2004-2008). More details are available at 
[Detecting Multiple Button Clicks in an Arduino Sketch](https://sigmdel.ca/michel/program/esp8266/arduino/switch_debouncing_en.html).


# Licence

The **BSD Zero Clause** ([SPDX](https://spdx.dev/): [0BSD](https://spdx.org/licenses/0BSD.html)) licence applies.






CHANGELOG.md

## Unreleased (development)


## Released


### 0.1.3

- Removed setCheckInterval from keywords.txt
- Updated version number in library.properties and library.json
- Set a default 0 value for the DEBUG_PUSH_BUTTON macro
- Removed Version section in README.md and 
- Added this file instead


### 0.1.2  96edee9

- Added button release debounce on long button presses.
- Removed the _checkInterval` and setCheckInterval. The functionality was not implemented.
- Improved the debugging functions in the mdPushButton class.
- Added the push_button_tune.ino sketch.
- Moved VERSION macro to mdPushButton.cpp from header file to avoid conflicts
- Fixed errors in README.md


### 0.1.1  9770957

- Changed time setters in mdPushButton
- Improved examples
- Updated README.md


### 0.1.0  46ba24a

- Initial commit to GitHub with the improvements listed below
- Added support for internal pull-down resistors with active HIGH buttons if available 
- Added support for active HIGH buttons 
- Added callback functions.


src/mdPushButton.cpp
/*
 * mdButton.cpp
 * 
 * A push button Arduino library capable of returning the number of
 * consecutive button presses made in quick succession or if the 
 * button was held down for a long time. 
 * 
 * Version 0.1.1, 2020-12-23.
 *
 * Michel Deslierres <sigmdel.ca/michel>
 * 
 */
 
 // SPDX-License-Identifier: 0BSD

#include "Arduino.h"
#include "mdPushButton.h"

#define VERSION 0x000101

// If active = LOW (the default) then the input must be pulled high most of the time
//   This can be done with the internal pullup (set pullup = true which is also the default value) 
//   or an external pull up resistor say 4.7K + connected to the  GPIO pin and Vcc
//
// If active = HIGH then the input must be pulled low most of the time
//   This can be done with the internal pulldown (set pullup = true which is also the default value) 
//   or an external pull down resistor say 4.7K + connected to the  GPIO pin and Vcc

mdPushButton::mdPushButton(uint8_t pin, uint8_t active, bool useInternalPullResistor) {
  _pin = pin;
  _active = active;

  int mode = INPUT;
  if (useInternalPullResistor) {
    if (active == LOW)
      mode = INPUT_PULLUP;
    else {
      // mode = INPUT_PULLDOWN where defined
      #if defined(INPUT_PULLDOWN) 
        mode = INPUT_PULLDOWN;
      #elif defined(ESP8266)
        if (pin == 16) mode = INPUT_PULLDOWN_16;
      #endif  
    }
  }  

  pinMode(_pin, mode);
  _buttonState = AWAIT_PRESS;
  version = VERSION;
}

//set time attributes

uint16_t  mdPushButton::_setAttrib(uint16_t* attrib, uint16_t value) {
  uint16_t time = *attrib;
  if (value < 0xFFFF)
    *attrib = value;
  return time;
}

uint16_t mdPushButton::setDebouncePressTime(uint16_t value) { return _setAttrib(&_debouncePressTime, value); }
uint16_t mdPushButton::setDebounceReleaseTime(uint16_t value) { return _setAttrib(&_debounceReleaseTime, value); }
uint16_t mdPushButton::setMultiClickTime(uint16_t value){ return _setAttrib(&_multiClickTime, value); }
uint16_t mdPushButton::setHoldTime(uint16_t value){ return _setAttrib(&_holdTime, value); }
uint16_t mdPushButton::setCheckInterval(uint16_t value) { return _setAttrib(&_checkInterval, value); }

void mdPushButton::OnButtonClicked( callback_int cb ){_OnClick1 = cb;}
void mdPushButton::OnButtonClicked( callback_int_int cb ){_OnClick2 = cb;}

int mdPushButton::status(void) {
  int stat = _update();
  if (stat) {
    if (_OnClick2) _OnClick2(_pin, stat);
    if (_OnClick1) _OnClick1(stat);
  }  
  return stat;
}  

#ifdef DEBUG_PUSH_BUTTON  

void mdPushButton::_debug(char * msg = NULL) {
  if (_lastState != _buttonState) {
    Serial.print("DBG pushButton: _pin ");
    Serial.print(_pin);
    Serial.print(" _clicks = ");
    Serial.print(_clicks);
    Serial.print(" _buttonState = ");
    switch(_buttonState) {
      case AWAIT_PRESS: Serial.print("AWAIT_PRESS"); break;
      case DEBOUNCE_PRESS: Serial.print("DEBOUNCE_PRESS");  break;
      case AWAIT_RELEASE: Serial.print("AWAIT_RELEASE");  break; 
      case DEBOUNCE_RELEASE: Serial.print("DEBOUNCE_RESLEASE");  break;
      case AWAIT_MULTI_PRESS: Serial.print("AWAIT_MULTI_PRESS");  break;
      default: Serial.print("*** UNKNOWN ***");
    };
    Serial.println(msg);
  }
  _lastState = _buttonState;  
}    
#endif

int mdPushButton::_update(void) {
  if (_buttonState == AWAIT_PRESS) {
    _clicks = 0;
    if (digitalRead(_pin) == _active) {
      _buttonState = DEBOUNCE_PRESS;
      _eventTime = millis();  
    }  
  } 
  
  else if (_buttonState == DEBOUNCE_PRESS) {
    if ((millis() - _eventTime) > _debouncePressTime) { 
      _buttonState = AWAIT_RELEASE;
      _eventTime = millis(); 
    }  
  } 
  
  else if (_buttonState == AWAIT_RELEASE) {
    //dbg: Serial.printf("%d ", digitalRead(_pin));
    if (!(digitalRead(_pin) == _active)) {
      if ((millis() - _eventTime) > _holdTime) {
        _buttonState = AWAIT_PRESS;
        #ifdef DEBUG_PUSH_BUTTON  
        _debug(" return -1");
        #endif
        return -1;
      }
      _buttonState = DEBOUNCE_RELEASE;
      _eventTime = millis();
    }
  } 

  else if (_buttonState == DEBOUNCE_RELEASE) {
    if ((millis() - _eventTime) > _debounceReleaseTime) { 
      _clicks += 1;
      _buttonState = AWAIT_MULTI_PRESS;
      _eventTime = millis(); 
    }  
  } 
  
  else {   // (buttonState == AWAIT_MULTI_PRESS) 
    if (digitalRead(_pin) == _active) {
      _buttonState = DEBOUNCE_PRESS;
      _eventTime = millis(); 
    } 
    else if ((millis() - _eventTime) > _multiClickTime) {
      _buttonState = AWAIT_PRESS;
        #ifdef DEBUG_PUSH_BUTTON  
        _debug(" return _clicks");
        #endif
      return _clicks;
    } 
  }
  #ifdef DEBUG_PUSH_BUTTON  
  _debug(" return 0");
  #endif  
  return 0; 
}



src/mdPushButton.h
/*
 * mdPushButton.h
 * 
 * A push button Arduino library capable of returning the number of
 * consecutive button presses made in quick succession or if the 
 * button was held down for a long time. 
 * 
 * Version 0.1.1, 2020-12-23.
 *
 * Michel Deslierres <sigmdel.ca/michel>
 * 
 * Version 0.1.1 
 *     - setXXXXTime(unint16_t value) functions now return the previous time value
 */
 
 // SPDX-License-Identifier: 0BSD

#ifndef MDPUSHBUTTON_H
#define MDPUSHBUTTON_H

#include "Arduino.h"

// May need to define this macro in Arduino IDE if a value != 0 is desired
//#define DEBUG_PUSH_BUTTON  = 0 - no debugging, 1 - printSetup() available, 2 adds state machine debugging

#define DEFAULT_DEBOUNCE_PRESS_TIME      15  // delay to debounce the make part of the signal
#define DEFAULT_DEBOUNCE_RELEASE_TIME    30  // delay to debounce the break part of the signal
#define DEFAULT_MULTI_CLICK_TIME        400  // if 0, does not check for multiple button clicks
#define DEFAULT_HOLD_TIME              2000  // minimum time of button press for mdButton.status() to return a -1 (long button press)
                                             // all times in milliseconds

// Callback type of handler such as buttonClicked(int clicks)
typedef void (*callback_int)(int);

// Callback type of handler such as buttonClicked(uint8_t buttonid, int clicks);
// the button id is the hopefully unique GPIO pin to which it is connected.
typedef void (*callback_int_int)(uint8_t, int);

class mdPushButton {
  public:
    mdPushButton(uint8_t pin, uint8_t active = LOW, bool useInternalPullResistor = true);  //constructor
   
    // Current library version
    int32_t version;

    // If value < 0xFFFF then sets the time attribute and returns its previous value.
    // If value == 0xFFFF returns the current time attribute
    uint16_t setDebouncePressTime(uint16_t value);
    uint16_t setDebounceReleaseTime(uint16_t value);
    uint16_t setMultiClickTime(uint16_t value);
    uint16_t setHoldTime(uint16_t value);

    // status, number of clicks since last update
    // -1 = button held, 0 = button not pressed, 1, 2, ... number of times button pressed
    int status();

    // Set callback function to be called when the button has been pressed   
    void OnButtonClicked(callback_int); // only one push button, or using separate click handlers for each button
    void OnButtonClicked(callback_int_int);  // when using one click handler for more than on push button

    #if (DEBUG_PUSH_BUTTON > 0)
    void printSetup(void);
  private:
    uint32_t _mode;
    #else
  private:
    #endif 
    uint8_t _pin;    
    uint8_t _active; 
    uint16_t _debouncePressTime   = DEFAULT_DEBOUNCE_PRESS_TIME; 
    uint16_t _debounceReleaseTime = DEFAULT_DEBOUNCE_RELEASE_TIME;
    uint16_t _multiClickTime      = DEFAULT_MULTI_CLICK_TIME;  
    uint16_t _holdTime            = DEFAULT_HOLD_TIME;
  
    callback_int _OnClick1;
    callback_int_int _OnClick2;

       // State variables 
    unsigned long lastButtonCheck_;
    long _eventTime; 
    enum buttonState_t { AWAIT_PRESS, DEBOUNCE_PRESS, AWAIT_RELEASE, DEBOUNCE_RELEASE, AWAIT_MULTI_PRESS } _buttonState;
    int _clicks;
    int _update(void);
    uint16_t _setAttrib(uint16_t* attrib, uint16_t value);

    // Debugging
    #if (DEBUG_PUSH_BUTTON > 1)
;   buttonState_t _lastState = AWAIT_MULTI_PRESS;
    void _debug(char * msg);
    #endif  
};

#endif
